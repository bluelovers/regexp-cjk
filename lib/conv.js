"use strict";
/**
 * Created by user on 2018/5/5/005.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports._word_zh_core2 = exports._word_zh_core = exports.zhTableAutoGreedyTable = void 0;
const index_1 = require("cjk-conv/lib/zh/table/index");
const util_1 = require("./util");
//console.log(cjkConv.zhTable.auto('é­¯'));
function zhTableAutoGreedyTable(s, options = {}) {
    if (true || options.greedyTable > 1) {
        options.safe = false;
    }
    options.greedyTable = options.greedyTable || true;
    return index_1.auto(s, options);
}
exports.zhTableAutoGreedyTable = zhTableAutoGreedyTable;
function _word_zh_core(search, skip, zhTableFn = index_1.auto, options = {}) {
    let opts;
    if (options.unsafe || true) {
        opts = {
            // @ts-ignore
            greedyTable: options.greedyTable,
            safe: false,
        };
    }
    return search.replace(util_1._re_cjk_conv('ug'), function (char) {
        if (skip && skip.indexOf(char) != -1) {
            return char;
        }
        let a = zhTableFn(char, opts);
        return a.length > 1 ? '[' + a.join('') + ']' : a[0];
    });
}
exports._word_zh_core = _word_zh_core;
function _word_zh_core2(search, skip, zhTableFn = index_1.auto, options = {}) {
    let opts;
    if (options.unsafe || true) {
        opts = {
            // @ts-ignore
            greedyTable: options.greedyTable,
            safe: false,
        };
    }
    return search.replace(util_1._re_cjk_conv('ug'), function (char) {
        if (skip && skip.indexOf(char) != -1) {
            return char;
        }
        let a = zhTableFn(char, opts);
        return a.join('');
    });
}
exports._word_zh_core2 = _word_zh_core2;
//# sourceMappingURL=conv.js.map