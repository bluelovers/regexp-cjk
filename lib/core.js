"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupParserEventEmitter = exports.coreHandler = exports.SymDefaults = exports.ParserEventEmitter = exports.ParserEventEmitterEvent = void 0;
const conv_1 = require("./conv");
const regexp_parser_event_1 = require("regexp-parser-event");
Object.defineProperty(exports, "ParserEventEmitter", { enumerable: true, get: function () { return regexp_parser_event_1.ParserEventEmitter; } });
const regexp_range_1 = __importDefault(require("regexp-range"));
const mergeOptions_1 = require("./mergeOptions");
const getSource_1 = __importDefault(require("./getSource"));
const index_1 = require("cjk-conv/lib/zh/table/index");
const plugin_1 = require("./plugin");
exports.SymDefaults = Symbol.for('zhRegExp.defaults');
function coreHandler(str, flags = null, options = {}, ...argv) {
    const opts = mergeOptions_1.getSettingOptions(str, flags, options, ...argv);
    let source;
    let hasFlags;
    ({ options, argv } = opts);
    options = mergeOptions_1.fixOptions(options);
    ({ source, hasFlags, flags } = getSource_1.default(opts));
    str = source;
    if (options.onCore) {
        let optsNew = options.onCore.reduce((a, setting) => {
            if (!setting.beforeStart) {
                return a;
            }
            return setting.beforeStart(a);
        }, {
            str,
            flags,
            options,
            argv,
            hasFlags,
        });
        ({ str, options, flags, argv, hasFlags } = optsNew);
    }
    if ((!options.disableZh || !options.disableLocalRange || options.on)) {
        let ev = regexp_parser_event_1.ParserEventEmitter.create(str, flags || '');
        const zhTableFn = options.zhTable || (options.greedyTable ? conv_1.zhTableAutoGreedyTable : index_1.auto);
        if (!options.disableZh) {
            ev.on("default" /* default */, function (ast) {
                plugin_1.astOldRaw(ast);
                let raw = conv_1._word_zh_core(ast.raw, options.skip, zhTableFn, options);
                if (ast.raw !== raw) {
                    ast.raw = raw;
                    ev.emit("change" /* change */, ast);
                }
            });
        }
        if (!options.disableLocalRange) {
            ev.on("class_range" /* class_range */, function (ast, ...argv) {
                let s = ast.min.raw;
                let e = ast.max.raw;
                let ret = regexp_range_1.default(s, e, {
                    createRegExpString: true,
                });
                if (ret) {
                    if (options.allowLocalRangeAutoZh) {
                        ret = conv_1._word_zh_core2(ret, options.skip, zhTableFn, options);
                    }
                    plugin_1.astOldRaw(ast);
                    if (ast.raw !== ret) {
                        ast.raw = ret;
                        ev.emit("change" /* change */, ast);
                    }
                }
            });
        }
        setupParserEventEmitter(ev, options);
        ev.resume();
        str = ev.getSource(!!options.debugChanged
            || !options.noUniqueClass
            || options.sortClass, options);
        flags = hasFlags ? flags : ev.flags;
    }
    if (options.onCore) {
        let optsNew = options.onCore.reduce((a, setting) => {
            if (!setting.afterStart) {
                return a;
            }
            return setting.afterStart(a);
        }, {
            str,
            flags,
            options,
            argv,
            hasFlags,
        });
        ({ str, options, flags, argv, hasFlags } = optsNew);
    }
    return {
        source: str,
        flags: flags || '',
        options: options,
    };
}
exports.coreHandler = coreHandler;
function setupParserEventEmitter(ev, options) {
    const onList = mergeOptions_1.fixOptions(options).on;
    if (onList) {
        onList
            .forEach((conf) => {
            Object
                .keys(conf)
                // @ts-ignore
                .forEach(function (event) {
                ev.on(event, conf[event]);
            });
        });
    }
    return ev;
}
exports.setupParserEventEmitter = setupParserEventEmitter;
exports.default = coreHandler;
//# sourceMappingURL=core.js.map