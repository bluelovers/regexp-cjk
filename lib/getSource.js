"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRegExpSourcePattern = exports.parseRegularExpressionString = void 0;
function parseRegularExpressionString(str) {
    let m = /^([\/#$%])(.+?)\1([a-z]*)$/.exec(str);
    if (m) {
        let [s, d, r, f] = m;
        return {
            source: typeof r !== 'undefined' ? r : '',
            flags: typeof f !== 'undefined' ? f : '',
            slash: s,
            input: str,
        };
    }
    return null;
}
exports.parseRegularExpressionString = parseRegularExpressionString;
function getRegExpSourcePattern(opts) {
    const { str, options } = opts;
    const { defaultFlags } = options;
    const hasFlags = typeof opts.flags == 'string';
    let source;
    let flags;
    if (str instanceof RegExp) {
        ({ source, flags } = str);
    }
    else if (typeof str === 'string') {
        let _do = true;
        if (options.parseRegularExpressionString) {
            let m = parseRegularExpressionString(str);
            if (m) {
                source = m.source;
                flags = m.flags;
                _do = false;
            }
        }
        if (_do) {
            source = str;
        }
    }
    else {
        throw new TypeError(`expected source is string or RegExp, but got '${str}', type: ${typeof str}`);
    }
    if (typeof source !== 'string') {
        throw new TypeError(`expected source is string, but got '${source}', type: ${typeof source}`);
    }
    flags = hasFlags ? opts.flags : flags;
    if (defaultFlags && (flags == null || flags === '')) {
        flags = defaultFlags;
    }
    return {
        source,
        flags,
        hasFlags,
    };
}
exports.getRegExpSourcePattern = getRegExpSourcePattern;
exports.default = getRegExpSourcePattern;
//# sourceMappingURL=getSource.js.map