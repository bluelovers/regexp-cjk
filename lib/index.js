"use strict";
/**
 * Created by user on 2018/5/3/003.
 *
 * 已廢棄 僅用於舊版相容
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports._word_zh_core = exports._word_zh = void 0;
const __1 = require("..");
const conv_1 = require("./conv");
Object.defineProperty(exports, "_word_zh_core", { enumerable: true, get: function () { return conv_1._word_zh_core; } });
function _word_zh(search, ret, flags = 'ig', skip) {
    let s;
    if (search instanceof RegExp) {
        s = new __1.zhRegExp(search, {
            skip,
        });
        flags = s.flags;
    }
    else {
        s = new __1.zhRegExp(search, flags, {
            skip,
        }).source;
    }
    return [s, ret, flags];
}
exports._word_zh = _word_zh;
//# sourceMappingURL=index.js.map