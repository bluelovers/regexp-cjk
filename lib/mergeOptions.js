"use strict";
/**
 * Created by user on 2019/6/15.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.fixOptions = exports.mergeOptions = exports.mergeOptions2 = exports.getSettingOptions = exports.customizer = void 0;
const lodash_1 = require("lodash");
const array_hyper_unique_1 = require("array-hyper-unique");
function customizer(objValue, srcValue) {
    if (lodash_1.isArray(objValue)) {
        return objValue.concat(srcValue);
    }
}
exports.customizer = customizer;
function getSettingOptions(str, flags = null, options = {}, ...argv) {
    if (flags !== null && typeof flags == 'object') {
        options = Object.assign({}, flags);
        flags = (options.flags || null);
    }
    if (typeof options == 'string') {
        options = {
            skip: options,
        };
    }
    if (typeof options.flags == 'string') {
        flags = options.flags;
    }
    return {
        str,
        flags: flags,
        options,
        argv,
    };
}
exports.getSettingOptions = getSettingOptions;
/**
 * for `zhRegExp.use` only
 */
function mergeOptions2(base = {}, ...opts) {
    let ret = mergeOptions(base, ...opts);
    if (typeof ret.flags === 'string') {
        ret.defaultFlags = ret.flags;
        delete ret.flags;
    }
    return ret;
}
exports.mergeOptions2 = mergeOptions2;
function mergeOptions(base = {}, ...opts) {
    let arr = [base || {}]
        .concat(opts)
        .filter(o => o);
    if (arr.length > 1) {
        base = lodash_1.mergeWith(...(arr.map(o => {
            return fixOptions(o);
            // @ts-ignore
        })), customizer);
    }
    return fixOptions(base);
}
exports.mergeOptions = mergeOptions;
/*
export function MergeDefaultOptions(target: typeof zhRegExp): typeof zhRegExp
{
    const zhRegExpNew = class zhRegExp extends target
    {
        constructor(...args: [any, ...any[]])
        {
            let { str, flags, options, argv } = getSettingOptions(...args);

            options = mergeOptions({}, zhRegExpNew[SymDefaults], options);

            super(str, flags, options, ...argv)
        }
    };

    return zhRegExpNew
}
 */
function fixOptions(options, removeEmptyOn) {
    if (options.on) {
        if (Array.isArray(options.on)) {
            array_hyper_unique_1.array_unique_overwrite(options.on);
        }
        else {
            options.on = [options.on];
        }
        options.on = options.on.filter(v => v);
        if (removeEmptyOn && !options.on.length) {
            delete options.on;
        }
    }
    if (removeEmptyOn && options.onCore) {
        if (!options.onCore.length) {
            delete options.onCore;
        }
    }
    // @ts-ignore
    return options;
}
exports.fixOptions = fixOptions;
exports.default = mergeOptions;
//# sourceMappingURL=mergeOptions.js.map