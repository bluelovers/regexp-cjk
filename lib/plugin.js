"use strict";
/**
 * Created by user on 2019/6/15.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.astNotChanged = exports.astOldRaw = void 0;
function astOldRaw(ast) {
    if (ast.old_raw == null) {
        ast.old_raw = ast.raw;
    }
    return ast.old_raw;
}
exports.astOldRaw = astOldRaw;
function astNotChanged(ast, notStrict) {
    astOldRaw(ast);
    if (notStrict) {
        return ast.old_raw === ast.raw;
    }
    return !ast.changed && ast.old_raw === ast.raw;
}
exports.astNotChanged = astNotChanged;
//# sourceMappingURL=plugin.js.map