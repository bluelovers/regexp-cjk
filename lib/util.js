"use strict";
/**
 * Created by user on 2018/5/5/005.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports._re_cjk_conv = void 0;
const cjk_conv_1 = require("regexp-helper/lib/cjk-conv");
/**
 * for regexp-cjk only
 */
function _re_cjk_conv(flags) {
    return cjk_conv_1._re_cjk_conv(flags, 'のと');
}
exports._re_cjk_conv = _re_cjk_conv;
//# sourceMappingURL=util.js.map