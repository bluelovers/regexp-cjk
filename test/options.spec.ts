import zhRegExp from '../index';
import { SymDefaults } from '../lib/core';

test(`defaultFlags`, () =>
{

	const flags = 'u';

	let ro = zhRegExp.use({
		flags,
	});

	let actual = new ro(/k/).flags;
	let expected = flags;

	expect(actual).toStrictEqual(expected);
	//expect(actual).toBeInstanceOf(Date);
	expect(actual).toMatchSnapshot();

	expect(new ro(/k/)).toMatchSnapshot();

});

test(`當 flags != '' 時 使用 flags`, () =>
{

	const flags = 'u';

	let ro = zhRegExp.use({
		flags,
	});

	let actual = new ro(/k/, 'g').flags;
	let expected = 'g';

	expect(actual).toStrictEqual(expected);
	//expect(actual).toBeInstanceOf(Date);
	expect(actual).toMatchSnapshot();

	expect(new ro(/k/, 'g')).toMatchSnapshot();

});

test(`當 flags = '' 時 使用 defaultFlags`, () =>
{

	const flags = 'u';

	let ro = zhRegExp.use({
		flags,
	});

	let r = new ro(/k/, '');

	let actual = r.flags;
	let expected = flags;

	expect(actual).toStrictEqual(expected);
	//expect(actual).toBeInstanceOf(Date);
	expect(actual).toMatchSnapshot();

	expect(r).toMatchSnapshot();
	expect(r).toBeInstanceOf(zhRegExp);

});

test(`當使用於 zhRegExp.use 內時 自動將 flags 轉換為 defaultFlags`, () =>
{

	const flags = 'u';

	let ro = zhRegExp.use({
		flags,
	});

	let actual = ro[SymDefaults];
	let expected = flags;

	expect(actual.defaultFlags).toStrictEqual(expected);
	expect(actual.flags).toBeUndefined();

	//expect(actual).toBeInstanceOf(Date);
	expect(actual).toMatchSnapshot();

});
