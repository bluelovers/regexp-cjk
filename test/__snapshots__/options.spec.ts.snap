// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultFlags 1`] = `"u"`;

exports[`defaultFlags 2`] = `/k/u`;

exports[`當 flags != '' 時 使用 flags 1`] = `"g"`;

exports[`當 flags != '' 時 使用 flags 2`] = `/k/g`;

exports[`當 flags = '' 時 使用 defaultFlags 1`] = `"u"`;

exports[`當 flags = '' 時 使用 defaultFlags 2`] = `/k/u`;

exports[`當使用於 zhRegExp.use 內時 自動將 flags 轉換為 defaultFlags 1`] = `
Object {
  "defaultFlags": "u",
}
`;

